server/service/software_installers.go

```diff
@@ -0,0 +1,144 @@
+package service
+
+import (
+	"context"
+	"fmt"
+	"mime/multipart"
+	"net/http"
+	"strconv"
+
+	"github.com/docker/go-units"
+	"github.com/fleetdm/fleet/v4/server/fleet"
+	"github.com/fleetdm/fleet/v4/server/ptr"
+)
+
+type uploadSoftwareInstallerRequest struct {
+	File              *multipart.FileHeader
+	TeamID            *uint
+	InstallScript     string
+	PreInstallQuery   string
+	PostInstallScript string
+}
+
+type uploadSoftwareInstallerResponse struct {
+	Err error `json:"error,omitempty"`
+}
+
+// TODO: We parse the whole body before running svc.authz.Authorize.
+// An authenticated but unauthorized user could abuse this.
+func (uploadSoftwareInstallerRequest) DecodeRequest(ctx context.Context, r *http.Request) (interface{}, error) {
+	decoded := uploadSoftwareInstallerRequest{}
+	err := r.ParseMultipartForm(512 * units.MiB)
+	if err != nil {
+		return nil, &fleet.BadRequestError{
+			Message:     "failed to parse multipart form",
+			InternalErr: err,
+		}
+	}
+
+	if r.MultipartForm.File["software"] == nil || len(r.MultipartForm.File["software"]) == 0 {
+		return nil, &fleet.BadRequestError{
+			Message:     "software multipart field is required",
+			InternalErr: err,
+		}
+	}
+
+	decoded.File = r.MultipartForm.File["software"][0]
+
+	if decoded.File.Size > 500*units.MiB {
+		// TODO: Should we try to assess the size earlier in the request processing (before parsing the form)?
```

Comment:
I don't think you can, I believe at that size it will be sent in chunks so you won't have a `Content-Length` header. (https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Transfer-Encoding#chunked)